src/python_testing/TC_CNET_4_23.py

```diff
@@ -0,0 +1,220 @@
+                TestStep(9, 'TH reads Networks attribute from the DUT PIXIT.CNET.ENDPOINT_THREAD endpoint. Verify that the Networks attribute list has an entry with the following values: network_id is the extended PAN ID of PIXIT.CNET.THREAD_1ST_OPERATIONALDATASET, Connected is of type bool and is TRUE. TH saves the index of the Networks list entry as `Userth_netidx`'),
+                TestStep(10, 'TH sends ArmFailSafe command to the DUT with ExpiryLengthSeconds set to 900| Verify that DUT sends ArmFailSafeResponse command to the TH'),
+                TestStep(11, 'TH sends RemoveNetwork Command to the DUT PIXIT.CNET.ENDPOINT_THREAD endpoint with network_id field set to the extended PAN ID of PIXIT.CNET.THREAD_1ST_OPERATIONALDATASET. Verify that DUT sends NetworkConfigResponse to command with the following response fields: NetworkingStatus is Success, NetworkIndex is `Userth_netidx`'),
+                TestStep(12, 'TH sends AddOrUpdateWiFiNetwork command to the DUT PIXIT.CNET.ENDPOINT_WIFI endpoint with SSID field set to PIXIT.CNET.WIFI_1ST_ACCESSPOINT_SSID, Credentials field set to PIXIT.CNET.WIFI_1ST_ACCESSPOINT_CREDENTIALS. Verify that DUT sends the NetworkConfigResponse command to the TH with the following response fields: NetworkingStatus is Success, DebugText is of type string with max length 512 or empty'),
+                TestStep(13, 'TH sends ConnectNetwork command to the DUT PIXIT.CNET.ENDPOINT_WIFI endpoint with network_id field set to PIXIT.CNET.WIFI_1ST_ACCESSPOINT_SSID'),
+                TestStep(14, 'TH sends the CommissioningComplete command to the DUT. Verify that DUT sends CommissioningCompleteResponse with the ErrorCode field set to OK'),
+                TestStep(15, 'TH reads Networks attribute from the DUT PIXIT.CNET.ENDPOINT_WIFI endpoint. Verify that the Networks attribute list has an entry with the following fields: network_id matches the bytes in PIXIT.CNET.WIFI_1ST_ACCESSPOINT_SSID, Connected is of type bool and is TRUE')]
+
+    def def_TC_CNET_4_23(self):
+        return '[TC-CNET-4.23] Verification for network interface switching between Wi-Fi and Thread [DUT-Server]'
+
+    def pics_TC_CNET_4_23(self):
+        return ["CNET.S.F00", "CNET.S.F01"]
+
+    # Override default timeout.
+    @property
+    def default_timeout(self) -> int:
+        return 200
+
+    async def SendConnectNetworkWithFailure(
+            self, networkID: int, endpoint: int):
+        with asserts.assert_raises(ChipStackError) as cm:
+            cmd = Clusters.NetworkCommissioning.Commands.ConnectNetwork(networkID=networkID)
+            await self.send_single_cmd(endpoint=endpoint, cmd=cmd)
+        asserts.assert_equal(cm.exception.err, kChipErrorTimeout, "Unexpected error while trying to send ConnectNetwork command")
+        logging.info("ConnectNetwork command timed out, not assert here as it is expected during network switching")
+
+    async def SendCommissioningCompleteWithRetry(self):
```

Comment:
I believe the reason you need the retries here is because you actually need to re-establish a new case session on the thread network. The retry here is waiting for the underlying session to error out and re-establish, but you can just force that to happen by expiring the session explicitly. This is done in some of the other CNET tests.